### ISSUE TITLE
Frozen files

### ISSUE THREAD

**Issue by slinnarsson** [linnarsson-lab/loompy#27]
Loom uses chunked, compressed storage in order to achieve reasonable performance along both axes of a layer. We use 64x64 chunks which we found a good tradeoff. 

However, if we knew that the matrix would only be accessed along the columns (say), then we could store it ordered by columns. This could potentially speed up access by an order of magnitude (or two). But, sadly, access along the other dimension would be *slowed down* by an order of magnitude (or two). For example, Joe Kington [did some benchmarking](http://geology.beer/2015/02/10/hdf-for-large-arrays/) on an  8GB dataset with three dimensions. With chunked storage, reading a slice along any dimension took 20 seconds. Without chunking, this dropped to 0.3 seconds along the favourable dimension, but increased to *3.5 hours* along the unfavourable one. So, this is why we use chunking.

Still, those 0.3 seconds are sure attractive. What if we would store each layer twice in the file, with one copy transposed? When the user slices along one of the dimensions, we would simply read from the appropriate underlying dataset. 

The problem is that we currently allow datasets to grow (by the addition of columns). Adding columns to the column-major dataset would be simple, but adding to the row-major one would be near-impossible.

Hence the proposal for "frozen" Loom files. A frozen file would work the same as a standard Loom file, but would be immutable. Once a file is frozen, it would be impossible to add columns (or rows) to any of the layers. But row and column attributes, and graphs, would still be mutable as before.

Here's an outline of the spec:

* Frozen loom files contain **three** copies of each layer: one chunked just like before, and two continuous, uncompressed. One of the continuous copies is transposed relative to the other. 

* Frozen files are created by freezing an existing file. Once frozen, the file becomes immutable, and cannot be thawed.

Benefits:

* Potentially 10 - 100x faster performance

* Current code would be completely unchanged except for a single call to ds.freeze()

* Backwards compatible, with same performance as before

Drawbacks:

* No compression, and duplication of data, so file size would grow about 5-10-fold. For example, a file currently compressed to 30% would contain the original layer (30% of the dataset size) plus two full-sized copies (200% of the dataset). Thus it would grow from 30% to 230%, or 7.5-fold.

So, basically, you would trade 10-100-fold speed for 5-10-fold file size. Might be a good deal in many cases.

**Comment by slinnarsson** [comment]
Turns out you get only approximately 2-fold increase in speed but 30-fold increase in size. That's not worth it. This was tested on a fast SSD, which is probably not much affected by random access patterns. But SSDs will soon be the norm, so it seems not worth the trouble to optimize for HDDs.

Anyway, it's implemented as `ds.freeze()` and seems to work ok, so I'll leave it in for now.

In the course of testing this, I found a performance bug in `ds.scan()`. It spent ~80% of the time slicing the graphs. But sliced graphs are typically useless in `scan()` anyway (since edges between windows are severed). I made an option `include_graphs` with default `False` and this sped up scanning by 4-5-fold.

I also found that using `int16` instead of `float32` speeds up scanning about 2-fold, and reduces file size 2-fold. That's not even a tradeoff, so we should use int16 as much as possible.

**Comment by JobLeonard** [comment]
> I also found that using int16 instead of float32 speeds up scanning about 2-fold, and reduces file size 2-fold. That's not even a tradeoff, so we should use int16 as much as possible.

This does indicate that the performance bottleneck is the data bandwidth - although it is not clear whether that is disk or RAM.

It's a shame that the [Blosc project](http://blosc.org/pages/blosc-in-depth/) is (currently) Python-specific and not easily integrated elsewhere, since it seems like a really good fit for this problem. [There is an HDF5 package](https://github.com/Blosc/hdf5-blosc) but it requires manual compilation to use, which is not very encouraging.

**Comment by slinnarsson** [comment]
Instead of freezing to dense uncompressed matrices, we could freeze to sparse matrices in CSR and CSC formats. They would be compressed to about 30% typically and should have fast access along one axis each. If the disk transfer speed is limiting, this could give us a 3-fold speedup, while causing only a ~3-fold increase in size.

Sparse matrices won't easily support numpy fancy indexing, so we'd have to implement that in loompy explicitly.

I'll try this when I find some time.

**Comment by slinnarsson** [comment]
Closing for lack of progress. All my testing has shown only modest gains with substantial increase in complexity (of implementation) and/or file size.
